### DISCUSSION TITLE
ViewModel’s leaked Flow collectors problem

### DISCUSSION THREAD

**Discussion by headsmanxxx** [orbit-mvi/orbit-mvi#213] [Q&A] [answered]
Hi, first of all, I find Orbit-MVI a wonderful library in any aspect. While watching samples I noticed two things using Flow collectors. One outside of repeatOnSubscriptionBlock and the other one inside of it. A similar methodology is used also in this tutorial from Land of Coding in Youtube: https://www.youtube.com/watch?v=T4NGT9DL3qw&t=78s

```
private fun requestStock() = intent(registerIdling = false) {
        stockRepository.stockDetails(itemName).collect {
            reduce {
                state.copy(stock = it)
            }
        }
    }
```

```
private fun requestStocks() = intent(registerIdling = false) {
        repeatOnSubscription {
            stockRepository.stockList().collect {
                reduce {
                    state.copy(stocks = it)
                }
            }
        }
    }
```

Is it a good idea to collect a flow inside of ViewModel taking into consideration the general approach for Flows: Don't use Flow.collect{} into your ViewModel? What happens in this case if requestStock() is public and this function will be triggered more than one time for whatever reason? Will this create new collectors all the time while all of them will remain in ViewModel until it is destroyed? Does repeatOnSubscription{} play any role here? Is it safe to call some different collectors in ViewModel and push their values into UiState? I'm very sorry if I have misunderstood something regarding the logic of the library for this topic.

**Comment by headsmanxxx** [comment]
@Rosomack or other guys?

**Comment by Rosomack** [comment, marked as answer]
A lot to unpack here :)

> Don't use Flow.collect{} into your ViewModel

I'd love to see the source! Collecting flows is fine as long as you practice good hygiene around the scoping.

> What happens in this case if requestStock() is public and this function will be triggered more than one time for whatever reason?

Every call will create a new coroutine collecting this flow. You as the developer are responsible for making sure it does not happen.

> while all of them will remain in ViewModel until it is destroyed?

Correct, until the ViewModel's scope is cancelled.

> Does repeatOnSubscription{} play any role here?

Not at all. `repeatOnSubscription` is an API that automatically cancels the given block when all subscribers to your container's `refCountStateFlows` (e.g. used via `collectAsState` in Compose) cancel. It also restarts the block when going from 0 to above zero subscribers.

Overall, it is a rare case  to expose an intent that starts a flow collector to be called from the outside.

Typically, you'd start collection inside the `onCreate` lambda of your container factory function. This is guaranteed to run only once during the ViewModel's lifecycle. A good practice is to also use `repeatOnSubscription` blocks to make sure that when your UI goes to the back stack and detaches, we pause listening to any flows we don't need.

```
class SomeViewModel(...): ViewModel(), ContainerHost<ExampleState, Nothing> {
    override val container = container<ExampleState, Nothing>(ExampleState()) {
        // Avoid launching individual intents here, use subintents. Subintents are simple suspend functions, but with orbit DSL available. 
        // That way you can `cancel` or `join` the `Job` returned by `runOnCreate` in your tests
        coroutineScope {
            launch { collectSomeIntent() }
            launch { collectAnotherIntent() }
        }
    }

    private suspend fun collectSomeIntent() = subIntent {
        repeatOnSubscription {
            someFlow.collect {
                ...
            }
        }
    }

    private suspend fun collectAnotherIntent() = subIntent {
        repeatOnSubscription {
            anotherFlow.collect {
                ...
            }
        }
    }
}
```
